addons/website_sale/models/product.py

```diff
@@ -371,11 +409,36 @@ def website_publish_button(self):
         self.ensure_one()
         return self.product_tmpl_id.website_publish_button()
 
+    @api.multi
+    def _get_images(self):
+        """Return a list of records implementing `image.mixin` to
+        display on the carousel on the website for this variant.
 
-class ProductImage(models.Model):
-    _name = 'product.image'
-    _description = 'Product Image'
+        This returns a list and not a recordset because the records might be
+        from different models (template, variant and image).
 
-    name = fields.Char('Name')
-    image = fields.Binary('Image', attachment=True)
-    product_tmpl_id = fields.Many2one('product.template', 'Related Product', copy=True)
+        It contains in this order: the main image of the variant (if set), the
+        Variant Extra Images, the main image of the template (conditionally) and
+        the Template Extra Images.
+        """
+        self.ensure_one()
+        variant_images = []
+        if self.image_raw_original:
```

Comment:
L426 if true: add image
L433 if false: add image

Reply:
These are opposite conditions because we always want to show the variant image

It's just showing it in a different position:

- We want to show it first if the variant image is set.
- We want to show it after the variant specific o2m when it's falling back to the template image, to always show variant images first and template images last.

Reply:
I updated the code to make it more clear and with only 1 if/else. Also used list comprehensions.